protocol/communities/manager.go

```diff
@@ -1370,65 +1371,144 @@ func (m *Manager) HandleCommunityEventsMessage(signer *ecdsa.PublicKey, message
+			if err != nil {
+				return updated, err
 			}
-		} else {
-			err := m.persistence.SaveRequestToJoin(requestToJoin)
+			err = m.persistence.SetRequestToJoinState(common.PubkeyToHex(pk), communityID, requestToJoin.State)
 			if err != nil {
-				return err
+				return updated, err
 			}
+			updated = true
+		}
+	} else {
+		err := m.persistence.SaveRequestToJoin(requestToJoin)
+		if err != nil {
+			return updated, err
 		}
-		return nil
 	}
+	return updated, nil
+}
 
-	for i := range community.config.EventsData.Events {
-		communityEvent := &community.config.EventsData.Events[i]
-		switch communityEvent.Type {
-		case protobuf.CommunityEvent_COMMUNITY_REQUEST_TO_JOIN_ACCEPT:
-			for signer, request := range communityEvent.AcceptedRequestsToJoin {
-				err := saveOrUpdateRequestToJoin(signer, request, RequestToJoinStateAccepted)
-				if err != nil {
-					return err
-				}
+func (m *Manager) handleCommunityEventRequestAccepted(community *Community, communityEvent *CommunityEvent) error {
+	requestToJoinState := RequestToJoinStateAccepted
+	if community.HasPermissionToSendCommunityEvents() {
+		// if we're an admin and we receive this admin event, we know the state is `pending`
+		requestToJoinState = RequestToJoinStateAcceptedPending
+	}
+
+	acceptedRequestsToJoin := make([]types.HexBytes, 0)
+
+	for signer, request := range communityEvent.AcceptedRequestsToJoin {
+		requestToJoin := &RequestToJoin{
+			PublicKey:   signer,
+			Clock:       request.Clock,
+			ENSName:     request.EnsName,
+			CommunityID: request.CommunityId,
+			State:       requestToJoinState,
+		}
+		requestToJoin.CalculateID()
+
+		if community.HasPermissionToSendCommunityEvents() {
+			existingRequestToJoin, err := m.persistence.GetRequestToJoin(requestToJoin.ID)
+			if err != nil && err != sql.ErrNoRows {
+				return err
+			}
+			if existingRequestToJoin.MarkedAsPendingByPrivilegedAccount() {
+				// the request is already in some pending state so we won't override it again
+				continue
 			}
+		}
 
-		case protobuf.CommunityEvent_COMMUNITY_REQUEST_TO_JOIN_REJECT:
-			for signer, request := range communityEvent.RejectedRequestsToJoin {
-				err := saveOrUpdateRequestToJoin(signer, request, RequestToJoinStateDeclined)
-				if err != nil {
-					return err
-				}
+		requestUpdated, err := m.saveOrUpdateRequestToJoin(signer, community.ID(), requestToJoin)
+		if err != nil {
+			return err
+		}
+
+		if requestUpdated {
+			// We only want to do this when `requestUpdated == true`, if the control node
+			// didn't know this request before, it means it just learned about it via this
+			// community event, which in turn means, we don't have any revealed addresses
+			// for this request yet, so we don't actually want to perform an
+			// accept action by the control node
+			acceptedRequestsToJoin = append(acceptedRequestsToJoin, requestToJoin.ID)
+		}
+	}
+	if community.IsControlNode() {
+		m.publish(&Subscription{AcceptedRequestsToJoin: acceptedRequestsToJoin})
```

Comment:
This event causes `Messenger` to `AcceptCommunityRequestToJoin()` for all accepted requests